src/irmin/tree.ml

```diff
@@ -722,42 +726,65 @@ module Make (P : Private.S) = struct
               | None -> of_t ()
               | Some _ as r -> Lwt.return r))
 
-    let list_of_map ?(offset = 0) ?length m : (step * elt) list =
-      let take_length seq =
-        match length with None -> List.of_seq seq | Some n -> Seq.take n seq
+    let seq_of_map ?(offset = 0) ?length m : (step * elt) Seq.t =
+      let take seq =
+        match length with None -> seq | Some n -> Seq.take n seq
       in
-      StepMap.to_seq m |> Seq.drop offset |> take_length
+      StepMap.to_seq m |> Seq.drop offset |> take
 
-    let list_of_value repo ?offset ?length v : (step * elt) list =
+    let seq_of_value repo ?offset ?length v : (step * elt) Seq.t =
       cnt.node_val_list <- cnt.node_val_list + 1;
-      let t = P.Node.Val.list ?offset ?length v in
-      List.fold_left
-        (fun acc (k, v) ->
+      let seq = P.Node.Val.seq ?offset ?length v in
+      Seq.map
+        (fun (k, v) ->
           match v with
           | `Node n ->
               let n = `Node (of_hash repo n) in
-              (k, n) :: acc
+              (k, n)
           | `Contents (c, m) ->
               let c = Contents.of_hash repo c in
-              (k, `Contents (c, m)) :: acc)
-        [] (List.rev t)
+              (k, `Contents (c, m)))
+        seq
 
-    let list ?offset ?length t : (step * elt) list or_error Lwt.t =
+    let seq ?offset ?length t : (step * elt) Seq.t or_error Lwt.t =
       match cached_map t with
-      | Some m -> ok (list_of_map ?offset ?length m)
+      | Some m -> ok (seq_of_map ?offset ?length m)
       | None -> (
           match t.v with
-          | Value (repo, n, None) -> ok (list_of_value ?offset ?length repo n)
+          | Value (repo, n, None) -> ok (seq_of_value ?offset ?length repo n)
           | Hash (repo, h) -> (
               value_of_hash t repo h >>= function
               | Error _ as e -> Lwt.return e
-              | Ok v -> ok (list_of_value ?offset ?length repo v))
+              | Ok v -> ok (seq_of_value ?offset ?length repo v))
           | _ -> (
               to_map t >>= function
               | Error _ as e -> Lwt.return e
-              | Ok m -> ok (list_of_map ?offset ?length m)))
+              | Ok m -> ok (seq_of_map ?offset ?length m)))
+
+    let seq_of_updates repo v updates =
+      let seq0 =
+        StepMap.to_seq updates
+        |> Seq.filter_map (function
+             | step, Add v -> Some (step, v)
+             | _, Remove -> None)
+      in
+      let seq1 =
+        P.Node.Val.seq v
+        |> Seq.filter (fun (step, _) -> not (StepMap.mem step updates))
+        |> Seq.map (fun (step, x) ->
+               let h = match x with `Node h -> h | `Contents (h, _) -> h in
+               let x = `Node (of_hash repo h) in
+               (step, x))
+      in
+      Seq.append seq0 seq1
```

Comment:
why are you looking at `v` at all here? I think we just care about the new possible nodes, which are only in `updates`.

Reply:
Good point, I'll change `seq_unordered` to `seq_newies` in order to only yield `Values`'s updates and `Map`'s elements